### PR TITLE
fix: OAuth code returned via redirect_uri

### PR DESCRIPTION
**Pull request by thomasslee97**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

This fixes the OAuth authentication procedure used by the Spreadsheet Gateway.

* **What is the current behavior?** (You can also link to an open issue here)

Currently the Spreadsheet Gateway relies on the user approving access to their Google Drive, then copying and pasting a code from a page displayed by Google into a form hosted in Sofie.

Since this was first implemented, Google has deprecated (and subsequently removed) support for this copy-paste method of entering access codes. Instead, it is required that Sofie provides a callback url which can be provided with the code as a URL parameter.

* **What is the new behavior (if this is a feature change)?**

Two new URL handlers have been added:
- `/devices/:deviceId/oauthResponse`
  - This handles the callback from the OAuth process and forwards the access code to the gateway
- `/devices/:deviceId/resetAuth`
  - The settings form for the OAuth flow now provides a button for resetting / clearing the auth settings from a peripheral device, offering a way to "sign out" from the Drive integration

* **Other information**:

There have also been a couple of QoL improvements made to the OAuth form, as well as removing references to "spreadsheet" to allow the form to be reused for other OAuth-based services.

**Status**

- [X] Code documentation for the relevant parts in the code have been added/updated by the PR author
- [X] The functionality has been tested by the PR author
- [ ] The functionality has been tested by NRK

### CODE REVIEW

File: meteor/server/api/peripheralDevice.ts [hunk 1]
```diff
@@ -609,6 +610,95 @@ PickerPOST.route('/devices/:deviceId/uploadCredentials', (params, req: IncomingM
...
+	} catch (e) {
+		res.statusCode = 500
+		content = e + ''
+		logger.error('Upload credentials failed: ' + e)
+	}
+
+	res.end(content)
+})
+
+PickerGET.route('/devices/:deviceId/resetAuth', (params, req: IncomingMessage, res: ServerResponse) => {
```
Comment (Julusian) [inline meteor/server/api/peripheralDevice.ts:667]: This should not be a GET. Perhaps a POST or maybe DELETE?

File: meteor/server/api/peripheralDevice.ts [hunk 2]
```diff
@@ -609,6 +610,95 @@ PickerPOST.route('/devices/:deviceId/uploadCredentials', (params, req: IncomingM
...
+
+		let accessToken = url.searchParams.get('code') || undefined
+		const scopes = url.searchParams.get('scope') || undefined
+
+		check(accessToken, String)
+		check(scopes, String)
+
+		accessToken = (accessToken + '').trim()
+		if (accessToken && accessToken.length > 5) {
+			PeripheralDeviceAPI.executeFunction(deviceId, 'receiveAuthToken', accessToken)
```
Comment (Julusian) [inline meteor/server/api/peripheralDevice.ts:642]: Is it intentional that we do the 302 redirect even if this executeFunction fails? A comment to state this and why would be good

File: meteor/server/api/peripheralDevice.ts [hunk 3]
```diff
@@ -609,6 +610,95 @@ PickerPOST.route('/devices/:deviceId/uploadCredentials', (params, req: IncomingM
...
+				.then(() => {
+					logger.info(`Sent auth token to device "${deviceId}"`)
+				})
+				.catch(logger.error)
+		}
+
+		const coreUrl = new URL(
+			`/settings/peripheralDevice/${deviceId}`,
```
Comment (Julusian) [inline meteor/server/api/peripheralDevice.ts:650]: I feel like it is possible to omit the `http://localhost:3000` from this redirect, but it has been a while

Comment (jstarpl) [reply meteor/server/api/peripheralDevice.ts:650]: Yes, it is perfectly safe to not provide protocol & host in the Location header.

**Review by Julusian — APPROVED** [review]

**Pull request merged**